proof/refine/ARM/InterruptAcc_R.thy

```diff
@@ -93,28 +93,357 @@ lemma work_units_and_irq_state_state_relationI [intro!]:
+         apply (corressimp corres: corres_machine_op)
+        apply (case_tac rv; clarsimp)
+        apply (rule corres_split_skip; corressimp corres: updateTimeStamp_corres)
+        apply (rule corres_split'[rotated 2, OF gets_sp getCurSc_sp])
+         apply (corressimp corres: getCurSc_corres)
+        apply (rule corres_split'[rotated 2, OF gets_sp getConsumedTime_sp])
+         apply (corressimp corres: getConsumedTime_corres)
+        apply (clarsimp simp: andM_def ifM_def bind_assoc)
+        apply (rule corres_split'[rotated 2, OF get_sc_active_sp scActive_sp])
+         apply (corressimp corres: scActive_corres)
+         apply (fastforce dest: valid_objs_valid_sched_context_size
+                          simp: cur_sc_tcb_def obj_at_def is_sc_obj_def sc_at_pred_n_def)
+        apply (clarsimp split: if_split)
+        apply (intro conjI impI)
+         apply (rule corres_guard_imp)
+           apply (rule corres_split[OF refillSufficient_corres]; simp)
+              apply (rule corres_split[OF isCurDomainExpired_corres])
+                apply (clarsimp simp: returnOk_def
+                               split: if_split)
+               apply wpsimp
+              apply (wpsimp simp: isCurDomainExpired_def)+
+          apply (prop_tac "is_active_sc (cur_sc s) s")
+           apply (clarsimp simp: obj_at_def vs_all_heap_simps active_sc_def)
+          apply (frule (1) active_sc_valid_refillsE)
+          apply (clarsimp simp: obj_at_def is_sc_obj_def sc_at_pred_n_def vs_all_heap_simps
+                                active_sc_def sc_valid_refills_def rr_valid_refills_def
+                         split: if_splits)
+         apply simp
+        apply corressimp
+       apply (fastforce intro: corres_returnOkTT)
+      apply (clarsimp split: if_split)
+     apply (clarsimp split: if_split)
+    apply (clarsimp simp: select_f_def mk_ef_def bind_def gets_def exs_valid_def get_def return_def
+                          wrap_ext_bool_det_ext_ext_def)
+   apply wpsimp
+   apply (clarsimp simp: select_f_def mk_ef_def bind_def gets_def get_def return_def
+                         work_units_limit_def wrap_ext_bool_det_ext_ext_def workUnitsLimit_def)
+  apply wpsimp
+  apply (clarsimp simp: select_f_def mk_ef_def bind_def gets_def exs_valid_def get_def return_def
+                        work_units_limit_def wrap_ext_bool_det_ext_ext_def workUnitsLimit_def )
+  done
+
+definition "updateTimeStamp_independent (P :: kernel_state \<Rightarrow> bool)
+  \<equiv> \<forall>f g s. P s \<longrightarrow> P (s\<lparr>ksCurTime := f (ksCurTime s), ksConsumedTime := g (ksConsumedTime s)\<rparr>)"
+
+lemma irq_state_independent_HI[intro!, simp]:
+  "\<lbrakk>\<And>s f g. P (s\<lparr>ksCurTime := f (ksCurTime s), ksConsumedTime := g (ksConsumedTime s)\<rparr>) = P s\<rbrakk>
+   \<Longrightarrow> updateTimeStamp_independent P"
+  by (simp add: updateTimeStamp_independent_def)
+
+definition "getCurrentTime_independent_D (P :: kernel_state \<Rightarrow> bool)
```

Comment:
This is more of a style issue, but is this `_D` suffix meant for something different from the existing `_H` as in `irq_state_independent_H`? If there is no difference, we probably want to stick to `_H`, or at least unify. Also, should `updateTimeStamp_independent` above also have the same suffix?

Reply:
I think I preferred using `_D` because it's for "design" rather than Haskell (or at least I guess the H is for Haskell). I might unify to `_D` unless there are objections. 

As for `updateTimeStamp_independent`, we don't need the suffix there because the equivalent for AInvs is `update_time_stamp_independent`, so there isn't a name clash like with the others

Reply:
I don’t have objections to `_D` itself, but there are `_H` in master, too... so, it’ll be at that level if we are unifying this.